replica/database.cc

```diff
@@ -1756,6 +1756,50 @@ future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db
     }
 }
 
+future<> database::apply(const std::vector<frozen_mutation>& muts, db::timeout_clock::time_point timeout) {
+    std::vector<commitlog_entry_writer> writers;
+    db::commitlog* cl = nullptr;
+    std::optional<shard_id> shard;
+
+    writers.reserve(muts.size());
+
+    for (auto i = 0; i < muts.size(); ++i) {
+        auto s = local_schema_registry().get(muts[i].schema_version());
+        auto&& cf = find_column_family(muts[i].column_family_id());
+
+        if (!cl) {
+            cl = cf.commitlog();
+        } else if (cl != cf.commitlog()) {
+            throw_with_backtrace<std::runtime_error>("Cannot apply atomically across commitlog domains");
+        }
+
+        auto m_shard = dht::shard_of(*s, dht::get_token(*s, muts[i].key()));
+        if (!shard) {
+            if (this_shard_id() != m_shard) {
+                throw_with_backtrace<std::runtime_error>(format("Must call apply() on the owning shard ({} != {})", this_shard_id(), m_shard));
+            }
+            shard = m_shard;
+        } else if (*shard != m_shard) {
+            throw_with_backtrace<std::runtime_error>("Cannot apply atomically across shards");
+        }
+
+        writers.emplace_back(s, muts[i], commitlog_entry_writer::force_sync::yes);
+    }
+
+    if (!cl) {
+        throw_with_backtrace<std::runtime_error>("Cannot apply atomically without commitlog");
+    }
+
+    std::vector<rp_handle> handles = co_await cl->add_entries(std::move(writers), timeout);
+
+    // FIXME: Memtable application is not atomic so reads may observe mutations partially applied until restart.
+    for (auto i = 0; i < muts.size(); ++i) {
+        auto&& cf = find_column_family(muts[i].column_family_id());
+        auto s = local_schema_registry().get(muts[i].schema_version());
+        co_await apply_in_memory(muts[i], s, std::move(handles[i]), timeout);
```

Comment:
The timeout is also a source of partial applies. I guess for the schema it doesn't matter, due to infinite timeout.